(Feature) Sell collectibles button

Agupane opened the pull request:
Closes #323 

It looks like this:
![image](https://user-images.githubusercontent.com/21086218/58129219-af4d7500-7bef-11e9-99a6-55a2bc37b4e5.png)

![image](https://user-images.githubusercontent.com/21086218/58129226-b4122900-7bef-11e9-880b-658d319ab6c1.png)

![image](https://user-images.githubusercontent.com/21086218/58129237-bb393700-7bef-11e9-8312-54f07730fe25.png)

Note: as I explained on my comment bellow, the "sell" button on the header should be styled on another pr, where all the styling of the headers is done

Agupane committed:
Creates new collectible list modal

Agupane committed:
Adds the sell collectible button
Adds a modal that lists the collectibles available to sell
The search and all the available collectibles to sell should be implemented

Agupane committed:
Now the collectible modal list uses all the user collectibles

Agupane committed:
Now the modals get closed if the user click outside them or use escape key

gabitoesmiapodo committed:
(add) collectibles list modal styles

Agupane committed:
Implements pure component to filter

Agupane committed:
Merge with styling commit
Adds filtering function to list component
Makes list component pure component

Agupane committed:
Removes wrong-committed constants file
Adds collectible list modal file

Agupane committed:
Removes warning from chevronRightIcon

Agupane committed:
Adds wrong-removed file

Agupane commented on src/components/erc721/collectibles/collectible_list_modal.tsx:
```diff
@@ -0,0 +1,178 @@
+    filterText: '',
+};
+
+// PureComponents only rerender if at least one state or prop value changes.
```
Based on [this code](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html)

Agupane commented on src/components/erc721/collectibles/collectible_list_modal.tsx:
```diff
@@ -0,0 +1,178 @@
+        return collectibles.filter((collectible: Collectible) => this._filterCollectibleByNameFn(collectible, name));
+    };
+
+    private readonly _filterCollectibleByNameFn = (collectible: Collectible, name: string): boolean => {
```
This could be exported on an utility function and used like [this PR](https://github.com/0xProject/0x-launch-kit-frontend/pull/418), thoughts @unjapones ?

Agupane replied:
It's merged now

Agupane commented on src/components/erc721/collectibles/collectibles_list.tsx:
```diff
@@ -48,6 +53,7 @@ export const CollectiblesList = (props: Props) => {
         <MainContainer>
             <FiltersMenu>
                 <Title>{props.title}</Title>
+                <Button onClick={props.toggleCollectibleListModal}>Sell collectibles</Button>
```
This button should be added on [this PR](https://github.com/0xProject/0x-launch-kit-frontend/pull/418), because in that PR the whole header is styled. Thoughts? @unjapones @gabitoesmiapodo

Agupane replied:
It's merged now

Agupane committed:
Updates modal title text

gabitoesmiapodo committed:
(fix) show loading while... weill, loading.

Agupane committed:
Merge branch 'development' of https://github.com/0xProject/0x-launch-kit-frontend into feature/collectiblesButton

# Conflicts:
#	src/components/erc721/collectibles/collectibles_list.tsx

Agupane committed:
Merge remote-tracking branch 'origin/feature/collectiblesButton' into feature/collectiblesButton

Agupane committed:
Refactors to use the filterable collectibles functions
Adds name as a new filter option

fvictorio commented:
If the search doesn't match anything, a spinner is shown instead of some message.

fvictorio commented:
Also: maybe we shouldn't show collectibles that are already for sale?

Agupane committed:
Shows empty component instead of loading spinner if the filter founds no results

Agupane committed:
Now the sell collectibles modal shown only the users collectibles that are available to sell (and not the ones that are already on sell)

Agupane commented:
It's done @fvictorio, now the modal display an "empty component" if no results were found. Also only displays the collectibles that are **not** already on sale.

fvictorio committed:
Replace ternary with null comparison

fvictorio commented on src/util/types.ts:
```diff
@@ -61,6 +61,7 @@ export interface UIState {
     readonly hasUnreadNotifications: boolean;
     readonly stepsModal: StepsModalState;
     readonly theme: Theme;
+    readonly isCollectibleListModalOpen: boolean;
```
Q: Does this need to be in the store?

Agupane replied:
It's not any more

fvictorio commented on src/components/erc721/common/inputSearch.tsx:
```diff
@@ -35,9 +42,15 @@ const SearchInput = styled.input`
+        return (
+            <SearchInput
+                {...restProps}
+                placeholder={placeHolder ? placeHolder : 'Search Cryptocards'}
```
I know it was used before, but I don't think we should use this default. Let's just make the placeholder mandatory. (Also, it's `placeholder`, not `placeHolder`).

Agupane replied:
Done

fvictorio commented on src/util/filterable_collectibles.ts:
```diff
@@ -22,12 +23,20 @@ const isCollectibleSoldInBasicSell = (collectible: Collectible): boolean => {
     return !isDutchAuction(collectible.order);
 };
 
+const filterCollectibleByNameFn = (collectible: Collectible, name: string): boolean => {
```
PP: Drop the `Fn` suffix.

Agupane replied:
Done

unjapones commented on src/components/erc721/collectibles/collectibles_list.tsx:
```diff
@@ -118,7 +122,9 @@ export class CollectiblesList extends React.Component<Props, State> {
-                    <ButtonStyled variant="quaternary">Sell collectibles</ButtonStyled>
+                    <ButtonStyled variant="quaternary" onClick={this.props.toggleCollectibleListModal}>
+                        Sell collectibles
+                    </ButtonStyled>
```
We can probably make this button a component itself so its parent doesn't need to map the dispatch and store properties.

This might be useful for the landing page of the erc721 app. We can do it in this PR or after #428 is merged (I don't want to add a dependency here).

Agupane committed:
Removes unused state on collectible list modal
Makes placeholder for search mandatory
Renames filterCollectibleByNameFn to filterCollectibleByName

Agupane committed:
Removes modal state from store, now the state is on the sell collectibles button

Agupane committed:
Removes modal state from store, now the state is on the sell collectibles button

unjapones committed:
Merge remote-tracking branch '0x/development' into feature/collectiblesButton

unjapones committed:
Fix compilation error.

unjapones committed:
Use indexOf() in filterCollectibleByName  function.

unjapones committed:
Simplify collectible_list_modal filter.

unjapones committed:
Fix erc721 InputSearch naming.

Simplify InputSearch Prop.

unjapones committed:
Add some types to collectible detail list/tile.

unjapones commented:
Overall, I believe it is good enough.

After we finish with the corresponding PRs, we can discuss the naming and component's folder structure in `erc721` because (at least for me) the names are a little bit confusing. We can probably go with something like:
```
collectibles_details/
    price_badge.tsx.
    collectibles_details_item_list.tsx
    collectibles_details_item_tile.tsx
    collectibles_details_list.tsx
collectibles_list/
    collectibles_card_list.tsx
    collectibles_list.tsx
    collectibles_sort_filter/
        collectibles_list_filter.tsx
        collectibles_list_sort.tsx
common/
    [...]
pages/
    [...]
modals/
    collectible_list_modal.tsx
    collectible_sell_modal.tes
    sell_collectibles_button.tsx
[...]
```

We'll see later, after we finish with the other PRs.

fvictorio committed:
Destructure props in the same line
